### ISSUE TITLE
Warnings of unused parameters from learner.cc

### ISSUE THREAD

**Issue by petecheng** [dmlc/xgboost#5717]
Environment: Ubuntu 18.04 LTS
My code is like this:
xgb_clf = xgb.XGBClassifier()
args = {'max_depth': p1, 'learning_rate': p2, 'booster': p3, 'n_jobs': p4,  'scale_pos_weight': p5, 'n_estimators': p6}
where p1-p6 is  predefined values.
xgb_clf.set_params(**args)

Then give me the warning:
WARNING: /workspace/src/learner.cc:480:
Parameters: { max_depth } might not be used.

  This may not be accurate due to some parameters are only used in language bindings but
  passed down to XGBoost core.  Or some parameters are not used but slip through this
  verification. Please open an issue if you find above cases.

But I am sure that max_depth is in the parameter list, so why this warning would occur?

**Comment by hcho3** [comment]
Don't use `set_params`. You should specify all parameters in the constructor:
```python
args = {'max_depth': p1, 'learning_rate': p2, 'booster': p3, 'n_jobs': p4, 'scale_pos_weight': p5, 'n_estimators': p6}
xgb_clf = xgb.XGBClassifier(**args)
```

**Comment by hcho3** [comment]
Oops my bad. Shouldn't have closed.

**Comment by petecheng** [comment]
> Don't use `set_params`. You should specify all parameters in the constructor:
> 
> ```python
> args = {'max_depth': p1, 'learning_rate': p2, 'booster': p3, 'n_jobs': p4,
>         'scale_pos_weight': p5, 'n_estimators': p6}
> xgb_clf = xgb.XGBClassifier(**args)
> ```

Thanks, and I will try. But in the documentation, it says the "set_paras" method would set the parameters of the estimator. Why does this not work?

**Comment by trivialfis** [comment]
@petecheng Which version of XGBoost are you using?  I just gave a tried on master branch:

``` python
import xgboost as xgb

xgb_clf = xgb.XGBClassifier()
args = {'max_depth': 10, 'learning_rate': 0.343, 'booster': 'gbtree',
        'n_jobs': 0, 'scale_pos_weight': 0.1, 'n_estimators': 100}
xgb_clf.set_params(**args)
```

No warning was emitted.

**Comment by petecheng** [comment]
> @petecheng Which version of XGBoost are you using? I just gave a try on master branch:
> 
> ```python
> import xgboost as xgb
> 
> xgb_clf = xgb.XGBClassifier()
> args = {'max_depth': 10, 'learning_rate': 0.343, 'booster': 'gbtree',
>         'n_jobs': 0, 'scale_pos_weight': 0.1, 'n_estimators': 100}
> xgb_clf.set_params(**args)
> ```
> 
> No warning was emitted.

The version is 1.1.0. And surprisingly I did not encounter this problem in another server (the same codes and running scripts, except that I specify a larger "n_jobs"(48, the cpu_cnt) where the warning happens), so I wonder if there are any environment checkings or other issues.

**Comment by trivialfis** [comment]
Maybe it's loading a wrong version of shared library?  could you try purging Xgboost from python site packages and reinstall Xgboost?

**Comment by petecheng** [comment]
> Maybe it's loading a wrong version of shared library? could you try purging Xgboost from python site packages and reinstall Xgboost?

OK, I will have a try. Thanks very much.

**Comment by trivialfis** [comment]
But scale pos weight is indeed not used, it's a regression only parameter.

**Comment by trivialfis** [comment]
Here's my complete script:
``` python
import xgboost as xgb
import numpy as np

rng = np.random.RandomState(1994)

kRows = 8
kCols = 4

X = rng.randn(kRows, kCols)
y = rng.randint(low=0, high=4, size=kRows)

xgb_clf = xgb.XGBClassifier()
args = {'max_depth': 10, 'learning_rate': 0.343, 'booster': 'gbtree',
        'n_jobs': 48, 'scale_pos_weight': 0.1, 'n_estimators': 100}
xgb_clf.set_params(**args)
xgb_clf.fit(X, y)
```

and the corresponding warning:
```
[20:09:12] WARNING: /home/fis/Workspace/XGBoost/xgboost/src/learner.cc:485: 
Parameters: { scale_pos_weight } might not be used.

  This may not be accurate due to some parameters are only used in language bindings but
  passed down to XGBoost core.  Or some parameters are not used but slip through this
  verification. Please open an issue if you find above cases.
```

**Comment by petecheng** [comment]
> But scale pos weight is indeed not used, it's a regression only parameter.

But  as the documentation said, scale_pos_weight is indeed a hyparameter of XGBClassifier, and as you posted before, when you give n_jobs 0, no warnings emitted.

**Comment by trivialfis** [comment]
I also tried different `n_jobs`, see above.

**Comment by trivialfis** [comment]
> scale_pos_weight is indeed a hyparameter of XGBClassifier

Blame inheritance.  ;-(  For good reference on parameters, see https://xgboost.readthedocs.io/en/latest/parameter.html .

**Comment by trivialfis** [comment]
My bad for being ambiguous.  It's for binary classification.  The warning shows up when you have multiple classes.

**Comment by trivialfis** [comment]
Will close in 3 days if no further updates for reproducing.

**Comment by jbibi1296** [comment]
I have the same error with XGBClassifier:
`Parameters: { min_samples_split,min_samples_leaf } might not be used. `
I used `verbosity = 0` and it helped take away the warning but now the models take 2-3 times longer to load compared to the older version of xgboost that I was running previously `xgboost 0.9.0`

**Comment by trivialfis** [comment]
@jbibi1296 I don't remember there are parameters called min sample split or min samples leaf.  Regarding the performance, could you provide an example script?

**Comment by trivialfis** [comment]
Just double checked, `min_samples_split,min_samples_leaf` are not valid, there are no such parameters in XGBoost.  Please open a separated issue for the performance issue.  Also we made some optimization on 1.1 release.  If you are using older version please upgrade.  Also we are spending a lots of time on `hist` and `gpu_hist` tree methods, please consider using it for better perf.

**Comment by fogloria** [comment]
> My bad for being ambiguous. It's for regression and binary classification, which is implemented as a logistic and treated differently than multi class classification (softmax). The warning shows up when you have multiple classes.

WARNING: /workspace/src/learner.cc:480: 
Parameters: { scale_pos_weight } might not be used.
so how do you fix this?

**Comment by trivialfis** [comment]
@fogloria There's nothing to fix.  For multi class classification the parameter is not valid.

**Comment by ihyaroshenko** [comment]
@trivialfis hi. Can you help, please? I have similar issue for multi class classification

```
import xgboost as xgb # version xgboost 1.3.3

XGB_PARAMETERS = {
            "booster": "gbtree", "verbosity": 1,
            "objective": "multi:softprob", "eval_metric": "merror",
            "importance_type": "gain",
            "learning_rate": 0.05, "n_estimators": 500,
            "max_depth": 5, "min_child_weight": 6, "gamma": 0,
            "subsample": 0.87, "colsample_bytree": 0.16,
            "reg_lambda": 100, "reg_alpha": 100,
            "sample_weight": classes_weights
        }
estimator = xgb.XGBClassifier()
estimator.set_params(**XGB_PARAMETERS)`
```
and the corresponding warning:

```
WARNING: /Users/travis/build/dmlc/xgboost/src/learner.cc:541: 
Parameters: { sample_weight } might not be used.

  This may not be accurate due to some parameters are only used in language bindings but
  passed down to XGBoost core.  Or some parameters are not used but slip through this
  verification. Please open an issue if you find above cases.

```
